modin/core/io/column_stores/parquet_dispatcher.py

```diff
@@ -542,7 +546,17 @@ def build_index(cls, dataset, partition_ids, index_columns):
         else:
             index_ids = [part_id[0][1] for part_id in partition_ids if len(part_id) > 0]
             index_objs = cls.materialize(index_ids)
-            complete_index = index_objs[0].append(index_objs[1:])
+            if range_index:
```

Comment:
If there are filters, but the index is not a range index, the `else` here is the right thing: just concatenate the indices. They may have repeated values, etc, which is fine.

A range index is the special case: we want the Modin dataframe to have a _single_ range, not a range that keeps restarting. i.e. if the partitions have index 0-9, 0-19, 0-29, we want our Modin dataframe to have 0-59. If I understand correctly, that is why this case exists:

https://github.com/modin-project/modin/blob/5b297a146e83d99a29fcfcaa3dc7dc20b60d02e4/modin/core/io/column_stores/parquet_dispatcher.py#L535-L538

by momentarily making a single Pandas dataframe, it can get what the full range would be, and not the ranges used within each partition.

So, I could have made this work with filters by passing the filters through to `to_pandas_dataframe` there. But that would actually require performing the filters, which is not a cheap operation like reading the index, and would have to be repeated again when the workers actually load in the data. So I figured having them do so, then using the materialized indices, was better. Let me know if I have missed something here.

Reply:
awesome explanation, thanks! maybe some of it could be left as in-code comment, too...